File: app/src/main/java/com/app/missednotificationsreminder/CustomApplicationBase.java
package com.app.missednotificationsreminder;

import android.app.Application;
import androidx.annotation.NonNull;
import android.util.Log;

import com.app.missednotificationsreminder.di.Injector;
import com.app.missednotificationsreminder.service.ReminderServiceJobCreator;
import com.app.missednotificationsreminder.ui.ActivityHierarchyServer;
import com.evernote.android.job.JobManager;
import com.jakewharton.threetenabp.AndroidThreeTen;
import com.jakewharton.u2020.data.LumberYard;

import javax.inject.Inject;

import dagger.ObjectGraph;
import timber.log.Timber;

/**
 * Custom application implementation to provide additional functionality such as dependency injection, leak inspection
 *
 * @author Eugene Popovich
 */
public abstract class CustomApplicationBase extends Application {

    private ObjectGraph mObjectGraph;

    @Inject ActivityHierarchyServer activityHierarchyServer;
    @Inject LumberYard lumberYard;

    public CustomApplicationBase() {
        // Initialize dependency injection graph
        mObjectGraph = ObjectGraph.create(getModules());
    }

    @Override public void onCreate() {
        super.onCreate();
        // leak inspection
        mObjectGraph.inject(this);
        // initialize logging
        if (BuildConfig.DEBUG) {
            Timber.plant(new Timber.DebugTree());
        } else {
            Timber.plant(new CrashReportingTree());
        }

        lumberYard.cleanUp();
        Timber.plant(lumberYard.tree());

        AndroidThreeTen.init(this);

        registerActivityLifecycleCallbacks(activityHierarchyServer);
        JobManager.create(this).addJobCreator(new ReminderServiceJobCreator());
    }

    abstract Object[] getModules();

    @Override
    public Object getSystemService(@NonNull String name) {
        if (Injector.matchesService(name)) {
            return mObjectGraph;
        }
        return super.getSystemService(name);
    }

    private static class CrashReportingTree extends Timber.DebugTree {
        @Override protected void log(int priority, String tag, String message, Throwable t) {
            if (priority == Log.VERBOSE || priority == Log.DEBUG) {
                // skip debug and verbose messages
                return;
            }
            super.log(priority, tag, message, t);
        }
    }
}

File: app/src/main/java/com/app/missednotificationsreminder/CustomApplicationBase.kt
package com.app.missednotificationsreminder

import android.content.Context
import android.util.Log
import androidx.multidex.MultiDex
import androidx.work.Configuration
import com.app.missednotificationsreminder.ui.ActivityHierarchyServer
import com.app.missednotificationsreminder.ui.activity.common.CommonActivityLifecycleCallback
import com.jakewharton.threetenabp.AndroidThreeTen
import com.jakewharton.u2020.data.LumberYard
import dagger.android.support.DaggerApplication
import timber.log.Timber
import timber.log.Timber.DebugTree
import javax.inject.Inject

/**
 * Custom application implementation to provide additional functionality such as dependency injection, leak inspection
 *
 * @author Eugene Popovich
 */
abstract class CustomApplicationBase : DaggerApplication(), Configuration.Provider {
    @Inject
    lateinit var activityHierarchyServer: ActivityHierarchyServer

    @Inject
    lateinit var lumberYard: LumberYard

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(base)
        MultiDex.install(this)
    }

    override fun onCreate() {
        AndroidThreeTen.init(this)
        super.onCreate()
        // initialize logging
        if (BuildConfig.DEBUG) {
            Timber.plant(DebugTree())
        } else {
            Timber.plant(CrashReportingTree())
        }
        lumberYard.cleanUp()
        Timber.plant(lumberYard.tree())
        registerActivityLifecycleCallbacks(activityHierarchyServer)
        registerActivityLifecycleCallbacks(CommonActivityLifecycleCallback())
    }

    override fun getWorkManagerConfiguration() =
            Configuration.Builder()
                    .setMinimumLoggingLevel(if (BuildConfig.DEBUG) Log.VERBOSE else Log.WARN)
                    .build()

    private class CrashReportingTree : DebugTree() {
        override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
            if (priority == Log.VERBOSE || priority == Log.DEBUG) {
                // skip debug and verbose messages
                return
            }
            super.log(priority, tag, message, t)
        }
    }
}